mirgecom/diffusion.py

```diff
@@ -430,6 +431,90 @@ def get_diffusion_flux(
             penalty_amount=penalty_amount)
 
 
+class RobinDiffusionBoundary(DiffusionBoundary):
+    r"""Robin boundary condition for the diffusion operator.
+
+    The non-homogeneous Robin boundary condition is a linear combination of
+    $u$ and its gradient $\nabla u$, given by
+
+    .. math::
+
+        (\alpha u - \kappa \nabla u \cdot \mathbf{\hat{n}})|_\Gamma =
+            \alpha u_{ref}.
+
+    where $u_{ref}$ is the reference value of $u$ for $x \to \infty$ and
+    $\alpha$ is the weight for $u$. The gradient weight $\kappa$ is the
+    conductivity (thermal) or the diffusivity (species).
+
+    The current implementation uses external data
+
+    .. math::
+
+        u^+ = u^-
+
+    when computing the boundary fluxes for $\nabla u$, and
+
+    .. math::
+
+        \nabla u\cdot\mathbf{\hat{n}} |_\Gamma =
+            \frac{\alpha}{\kappa}(u_{ref} - u^-)
+
+    when computing the boundary fluxes for $\nabla \cdot (\kappa \nabla u)$.
+
+    .. automethod:: __init__
+    .. automethod:: get_grad_flux
+    .. automethod:: get_diffusion_flux
+    """
+
+    def __init__(self, value, alpha):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        value: float or meshmode.dof_array.DOFArray
+            the reference value(s) of $u$ along the boundary
+        alpha: float or meshmode.dof_array.DOFArray
+            the weight for the variable $u$ at the boundary
+        """
+        self.value = value
+        self.alpha = alpha
+
+    def get_grad_flux(
+            self, dcoll, dd_bdry, kappa_minus, u_minus, *,
+            numerical_flux_func=grad_facial_flux_weighted):  # noqa: D102
+        actx = u_minus.array_context
+        kappa_tpair = TracePair(dd_bdry,
+            interior=kappa_minus,
+            exterior=kappa_minus)
+        u_tpair = TracePair(dd_bdry,
+            interior=u_minus,
+            exterior=u_minus)
+        normal = actx.thaw(dcoll.normal(dd_bdry))
+        return numerical_flux_func(kappa_tpair, u_tpair, normal)
+
+    def get_diffusion_flux(
+            self, dcoll, dd_bdry, kappa_minus, u_minus, grad_u_minus,
+            lengthscales_minus, *, penalty_amount=None,
+            numerical_flux_func=diffusion_facial_flux_harmonic):  # noqa: D102
+        actx = u_minus.array_context
+        kappa_tpair = TracePair(dd_bdry,
+            interior=kappa_minus,
+            exterior=kappa_minus)
+        u_tpair = TracePair(dd_bdry,
+            interior=u_minus,
+            exterior=u_minus)
+        normal = actx.thaw(dcoll.normal(dd_bdry))
+        grad_u_tpair = TracePair(dd_bdry,
+            interior=grad_u_minus,
+            exterior=self.alpha*(self.value - u_minus)/kappa_minus * normal)
```

Comment:
```suggestion
        dudn_bc = self.alpha * (self.value - u_minus)/kappa_minus
        grad_u_tpair = TracePair(dd_bdry,
            interior=grad_u_minus,
            exterior=grad_u_minus + 2 * (dudn_bc - np.dot(grad_u_minus, normal)) * normal)
```

I think...

Reply:
Implemented the suggested expression but in a different form. See https://github.com/illinois-ceesd/mirgecom/pull/993/commits/7083161b637591e9f9163b13b370381cdbd60967

Reply:
Does that version work with orthotropic kappa, since I think in that case `dot(kappa*grad_u, normal) != dot(kappa*grad_u_normal, normal)` (in other words, kappa changes the vector direction instead of just scaling it)?

Reply:
Agreed that it is different. However, we only have orthotropic/anisotropic diffusivities for heat, not for species (which is the main point of this BC in our cases).

I will add a comment wrt anisotropy, but I wouldn't put effort now in a case we will likely never need.

Reply:
Tempted to say the original suggested version would work though, since it preserves the perpendicular components of `grad_u` (instead of zeroing them out), and just reflects the normal component... agree/disagree?

Reply:
I see my mistake now. Agreed and kept the original suggestion in https://github.com/illinois-ceesd/mirgecom/pull/993/commits/33bf186853a614a2a3703da8fe7a64034ad0f8b2